libjcat/jcat-self-test.c

```diff
@@ -2613,9 +2613,308 @@ TestVerifyInclusionProof(void)
+		} else if (inmemoryTreeNodeCount(tree, level + 1) == (first_node >> 1) + 1) {
+			GPtrArray *next_level = g_ptr_array_index(tree->tree, level + 1);
+			g_ptr_array_remove_index(next_level, next_level->len - 1);
+		}
+		for (gint64 j = first_node & ~(gint64)1; j < last_node; j += 2) {
+			g_ptr_array_add(
+			    g_ptr_array_index(tree->tree, level + 1),
+			    jcat_rfc6962_hash_children(
+				g_ptr_array_index((GPtrArray *)g_ptr_array_index(tree->tree, level),
+						  j),
+				g_ptr_array_index((GPtrArray *)g_ptr_array_index(tree->tree, level),
+						  j + 1)));
+		}
+		if ((last_node & 1) == 0) {
+			g_ptr_array_add(g_ptr_array_index(tree->tree, level + 1),
+					g_byte_array_ref(g_ptr_array_index(
+					    (GPtrArray *)g_ptr_array_index(tree->tree, level),
+					    last_node)));
+		}
+	}
+	tree->leavesProcessed = snapshot;
+	return inmemoryTreeRoot(tree);
+}
+
+static GByteArray *
+inmemoryTreeRecomputePastSnapshot(inmemoryTree *tree,
+				  gint64 snapshot,
+				  gint64 node_level,
+				  GByteArray **node)
+{
+	gint64 level = 0;
+	gint64 last_node = snapshot - 1;
+	g_autoptr(GByteArray) subtree_root = NULL;
+	if (snapshot == tree->leavesProcessed) {
+		if (node != NULL && tree->tree->len > node_level) {
+			GPtrArray *this_level = g_ptr_array_index(tree->tree, node_level);
+			if (*node != NULL)
+				g_byte_array_unref(*node);
+			if (node_level > 0) {
+				*node = g_byte_array_ref(
+				    g_ptr_array_index(this_level, this_level->len - 1));
+			} else {
+				*node = g_byte_array_ref(g_ptr_array_index(this_level, last_node));
+			}
+		}
+		return inmemoryTreeRoot(tree);
+	}
+
+	g_assert_cmpint(snapshot, <, tree->leavesProcessed);
+
+	while ((last_node & 1) == 1) {
+		if (node != NULL && node_level == level) {
+			GPtrArray *this_level = g_ptr_array_index(tree->tree, level);
+			if (*node != NULL)
+				g_byte_array_unref(*node);
+			*node = g_byte_array_ref(g_ptr_array_index(this_level, last_node));
+		}
+		last_node >>= 1;
+		level++;
+	}
+
+	subtree_root = g_byte_array_ref(
+	    g_ptr_array_index((GPtrArray *)g_ptr_array_index(tree->tree, level), last_node));
+
+	if (node != NULL && node_level == level) {
+		if (*node != NULL)
+			g_byte_array_unref(*node);
+		*node = g_byte_array_ref(subtree_root);
+	}
+
+	while (last_node != 0) {
+		if ((last_node & 1) == 1) {
+			GPtrArray *this_level = g_ptr_array_index(tree->tree, level);
+			GByteArray *new_subtree_root =
+			    jcat_rfc6962_hash_children(g_ptr_array_index(this_level, last_node - 1),
+						       subtree_root);
+			g_byte_array_unref(subtree_root);
+			subtree_root = new_subtree_root;
+		}
+
+		last_node >>= 1;
+		level++;
+		if (node != NULL && node_level == level) {
+			if (*node != NULL)
+				g_byte_array_unref(*node);
+			*node = g_byte_array_ref(subtree_root);
+		}
+	}
+	return subtree_root;
+}
+
+static GByteArray *
+inmemoryTreeRootAtSnapshot(inmemoryTree *tree, gint64 snapshot)
+{
+	if (snapshot == 0) {
+		return g_bytes_unref_to_array(sha256EmptyTreeHash());
+	}
```

Comment:
just stylisticly, you don't need {} for single line statements like this.

Reply:
Done.